test: prevent memory leak in DB Console tests

sjbarag opened the pull request:
It appears there was a memory leak in the tests for pkg/ui/workspaces/
db-console, but not in typical way for a set of JavaScript tests. Rather
than v8's JavaScript heap running out of memory (typically solved by
increasing --max-old-space-size [1]), the v8 JS virtual machine *itself*
seems to be leaking memory. This was only observable by polling `free`
(or watching `htop` and friends), and presented as ever-increasing
memory pressure on the host OS. The systems that run these tests
in CI don't have any swap space available, so this memory pressure
eventually spills over and the entire machine becomes largely
unresponsive [2]. Eventually -- after one hour, thanks to
test_size=enormous [3] -- Bazel force-kills the node binary and reports
the tests as failed.

There's a few references to this behavior on the Internet [4], and the
general consensus is that this started with node 16.11.0. Newer versions
of node, including the current LTS 18.15.0, appear to exhibit the same
leak. New per-worker memory limit options were added to Jest in its
most-recent major version [5], but the upgrade path from our current
27.x to 29.x is non-trivial. Forcing each jest worker to perform a GC
pass after each suite run seems to solve the issue on older versions
well enough, and should allow these tests to run to again completion
without hanging.

[1] https://nodejs.org/docs/latest-v16.x/api/cli.html#--max-old-space-sizesize-in-megabytes
[2] This is reproducible when SSH'd into a remote GCP or AWS instance
    running both x86_64 and aarch64 (where the SSH connection drops),
    but similar behavior seems to happen in TeamCity.
    Aside: It's unclear why neither Bazel nor node were killed via
    some OOM watchdog. Perhaps there simply isn't one configured on
    these machines, or perhaps v8 stops JS execution while waiting for
    additional memory to be allocated that the OS can't provide? Knowing
    why unfortunately doesn't fix the leak.
[3] https://bazel.build/reference/be/common-definitions#test.size
[4] https://github.com/facebook/jest/issues/11956
[5] https://jestjs.io/docs/29.0/configuration#workeridlememorylimit-numberstring

Fixes: https://github.com/cockroachdb/cockroach/issues/97179

Release note (build change): Running './dev ui test' (or
'bazel test //pkg/ui/workspaces/db-console:jest') now uses less memory.

sjbarag committed:
test: prevent memory leak in DB Console tests

It appears there was a memory leak in the tests for pkg/ui/workspaces/
db-console, but not in typical way for a set of JavaScript tests. Rather
than v8's JavaScript heap running out of memory (typically solved by
increasing --max-old-space-size [1]), the v8 JS virtual machine *itself*
seems to be leaking memory. This was only observable by polling `free`
(or watching `htop` and friends), and presented as ever-increasing
memory pressure on the host OS. The systems that run these tests
in CI don't have any swap space available, so this memory pressure
eventually spills over and the entire machine becomes largely
unresponsive [2]. Eventually -- after one hour, thanks to
test_size=enormous [3] -- Bazel force-kills the node binary and reports
the tests as failed.

There's a few references to this behavior on the Internet [4], and the
general consensus is that this started with node 16.11.0. Newer versions
of node, including the current LTS 18.15.0, appear to exhibit the same
leak. New per-worker memory limit options were added to Jest in its
most-recent major version [5], but the upgrade path from our current
27.x to 29.x is non-trivial. Forcing each jest worker to perform a GC
pass after each suite run seems to solve the issue on older versions
well enough, and should allow these tests to run to again completion
without hanging.

[1] https://nodejs.org/docs/latest-v16.x/api/cli.html#--max-old-space-sizesize-in-megabytes
[2] This is reproducible when SSH'd into a remote GCP or AWS instance
    running both x86_64 and aarch64 (where the SSH connection drops),
    but similar behavior seems to happen in TeamCity.

    Aside: It's unclear why neither Bazel nor node were killed via
    some OOM watchdog. Perhaps there simply isn't one configured on
    these machines, or perhaps v8 stops JS execution while waiting for
    additional memory to be allocated that the OS can't provide? Knowing
    why unfortunately doesn't fix the leak.
[3] https://bazel.build/reference/be/common-definitions#test.size
[4] https://github.com/facebook/jest/issues/11956
[5] https://jestjs.io/docs/29.0/configuration#workeridlememorylimit-numberstring

Fixes: #97179

Release note (build change): Running './dev ui test' (or
'bazel test //pkg/ui/workspaces/db-console:jest') now uses less memory.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/98880)
<!-- Reviewable:end -->

sjbarag committed:
CI: re-enable //pkg/ui/workspaces/db-console:jest under arm64

This reverts the part of commit 0f743afc584b1837c9e2f1bb9683eb58f67166ef
from https://github.com/cockroachdb/cockroach/pull/98694.

Release note: None
Epic: none

sjbarag commented:
x86_64 completed: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_Tests_UiTest/9117298?showRootCauses=true&expandBuildChangesSection=true

arm64 is still running at 481s: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UiTest/9117438?showRootCauses=true&expandBuildChangesSection=true

sjbarag commented:
bors r=healthy-pod p=99

healthy-pod commented:
bors r+ p=99

sjbarag commented:
bors r+ p=99

renatolabs commented:
bors r+ p=99
